worker/raft/rafttransport/worker.go

```diff
@@ -0,0 +1,291 @@
+	}
+	if config.Hub == nil {
+		return errors.NotValidf("nil Hub")
+	}
+	if config.Mux == nil {
+		return errors.NotValidf("nil Mux")
+	}
+	if config.Path == "" {
+		return errors.NotValidf("empty Path")
+	}
+	if config.Tag == nil {
+		return errors.NotValidf("nil Tag")
+	}
+	if config.TLSConfig == nil {
+		return errors.NotValidf("nil TLSConfig")
+	}
+	return nil
+}
+
+// NewWorker returns a new apiserver-based raft transport worker,
+// with the given configuration. The worker itself implements
+// raft.Transport.
+func NewWorker(config Config) (*Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	apiPorts := config.APIInfo.Ports()
+	if n := len(apiPorts); n != 1 {
+		return nil, errors.Errorf("api.Info has %d unique ports, expected 1", n)
+	}
+
+	w := &Worker{
+		config:       config,
+		connections:  make(chan net.Conn),
+		dialRequests: make(chan dialRequest),
+		apiPort:      apiPorts[0],
+	}
+
+	const logPrefix = "[transport] "
+	logWriter := &raftutil.LoggoWriter{logger, loggo.DEBUG}
+	logLogger := log.New(logWriter, logPrefix, 0)
+	transport := raft.NewNetworkTransportWithConfig(&raft.NetworkTransportConfig{
+		Logger:  logLogger,
+		MaxPool: maxPoolSize,
+		Stream: newStreamLayer(config.Tag, config.Hub, w.connections, &Dialer{
+			APIInfo: config.APIInfo,
+			DialRaw: w.dialRaw,
+			Path:    config.Path,
+		}),
+		Timeout: config.Timeout,
+	})
+	w.Transport = transport
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: func() error {
+			defer transport.Close()
+			return w.loop()
+		},
+	}); err != nil {
+		transport.Close()
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+// Worker is a worker that manages a raft.Transport.
+type Worker struct {
+	raft.Transport
+
+	catacomb     catacomb.Catacomb
+	config       Config
+	connections  chan net.Conn
+	dialRequests chan dialRequest
+	tlsConfig    *tls.Config
+	apiPort      int
+}
+
+type dialRequest struct {
+	ctx     context.Context
+	address string
+	result  chan<- dialResult
+}
+
+type dialResult struct {
+	conn net.Conn
+	err  error
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *Worker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *Worker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+// dialRaw dials a new TLS connection to the controller identified
+// by the given address. The address is expected to be the stringified
+// tag of a controller machine agent. The resulting connection is
+// appropriate for use as Dialer.DialRaw.
+func (w *Worker) dialRaw(address raft.ServerAddress, timeout time.Duration) (net.Conn, error) {
+	// Give precedence to the worker dying.
+	select {
+	case <-w.catacomb.Dying():
+		return nil, w.errDialWorkerStopped()
+	default:
+	}
+
+	ctx := context.Background()
+	if timeout != 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
```

Comment:
presumably the Cancel is to cancel timing out?
Or is it to cancel the connection?
(I'm presuming the former, as a way to disable the timeout if this function returns successfully)

Reply:
Mostly to cancel the dial timeout timer. It's a pattern for all context usage: if you have get a cancel function, you defer it in the current function.

From the docs:

> The WithCancel, WithDeadline, and WithTimeout functions take a Context (the parent) and return a derived Context (the child) and a CancelFunc. Calling the CancelFunc cancels the child and its children, removes the parent's reference to the child, and stops any associated timers. Failing to call the CancelFunc leaks the child and its children until the parent is canceled or the timer fires. The go vet tool checks that CancelFuncs are used on all control-flow paths.